Fix bad assert when a DeferredWorkTimer ticket is cancelled as its global object is visiting children

kmiller68 opened the pull request:
#### 9bc2ec09cdb0f556906527c7d69603037f5ec673
<pre>
Fix bad assert when a DeferredWorkTimer ticket is cancelled as its global object is visiting children
<a href="https://bugs.webkit.org/show_bug.cgi?id=282125">https://bugs.webkit.org/show_bug.cgi?id=282125</a>
<a href="https://rdar.apple.com/138675242">rdar://138675242</a>

Reviewed by Yijia Huang and Mark Lam.

The isCancelled check in JSGlobalObject&apos;s visitChildren is just an optimization since the mutator is
running while visiting. Thus the ticket could get cancelled between the check and the dependency fetch.
Thus we should just relax the assertion in this case.

* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
(JSC::DeferredWorkTimer::TicketData::dependencies):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::visitChildrenImpl):

Canonical link: <a href="https://commits.webkit.org/285781@main">https://commits.webkit.org/285781@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86e7b3af86ee6d6ea83af438ae0a9cb3c4b44dcb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73795 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53224 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26606 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78117 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25033 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62357 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1009 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16417 "2 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76862 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63500 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/38446 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44983 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20988 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23366 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66932 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/66551 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21335 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79684 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73053 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1112 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/561 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1255 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63511 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16226 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9533 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7720 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/94834 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1076 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20852 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1105 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1092 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1111 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (approved):
r=me

hyjorc1 commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
```diff
@@ -62,7 +62,7 @@ class DeferredWorkTimer final : public JSRunLoopTimer {
         inline VM& vm();
         JSObject* target();
         inline bool hasValidTarget() const;
```
If we don't clear `m_dependencies` in `DeferredWorkTimer::TicketData::cancel()`, then we don't need this `hasValidTarget()` any more.

hyjorc1 replied:
By doing that, could you please add an assertion in `DeferredWorkTimer::TicketData::TicketData` or `DeferredWorkTimer::TicketData::create`?
```
ASSERT(!m_dependencies.isEmpty());
```
This is because the last one should always be `target` in `DeferredWorkTimer::addPendingWork`.
```
    dependencies.append(Weak<JSCell>(target));

    auto ticketData = TicketData::create(type, globalObject, scriptExecutionOwner, WTFMove(dependencies));
```

hyjorc1 reviewed (approved):
r=me with comments

kmiller68 committed:
Fix bad assert when a DeferredWorkTimer ticket is cancelled as its global object is visiting children
https://bugs.webkit.org/show_bug.cgi?id=282125
rdar://138675242

Reviewed by Yijia Huang and Mark Lam.

The isCancelled check in JSGlobalObject's visitChildren is just an optimization since the mutator is
running while visiting. Thus the ticket could get cancelled between the check and the dependency fetch.
Thus we should just relax the assertion in this case.

* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
(JSC::DeferredWorkTimer::TicketData::dependencies):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::visitChildrenImpl):

Canonical link: https://commits.webkit.org/285781@main
